core/primitives/src/stateless_validation/contract_distribution.rs

```diff
@@ -0,0 +1,270 @@
+use borsh::{BorshDeserialize, BorshSerialize};
+use bytesize::ByteSize;
+use near_crypto::Signature;
+use near_primitives_core::{
+    hash::CryptoHash,
+    types::{BlockHeight, ShardId},
+};
+use near_schema_checker_lib::ProtocolSchema;
+
+use crate::{
+    types::EpochId, utils::compression::CompressedData, validator_signer::ValidatorSigner,
+};
+
+use super::{ChunkProductionKey, SignatureDifferentiator};
+
+/// Contains contracts (as code-hashes) accessed during the application of a chunk.
+/// This is used by the chunk producer to let the chunk validators know about which contracts
+/// are needed for validating a witness, so that the chunk validators can request missing code.
+#[derive(Debug, Clone, PartialEq, Eq, BorshSerialize, BorshDeserialize, ProtocolSchema)]
+pub enum ChunkContractAccesses {
+    V1(ChunkContractAccessesV1),
+}
+
+impl ChunkContractAccesses {
+    pub fn new(
+        next_chunk: ChunkProductionKey,
+        contracts: Vec<CodeHash>,
+        signer: &ValidatorSigner,
+    ) -> Self {
+        Self::V1(ChunkContractAccessesV1::new(next_chunk, contracts, signer))
+    }
+
+    pub fn contracts(&self) -> &Vec<CodeHash> {
+        match self {
+            Self::V1(accesses) => &accesses.inner.contracts,
+        }
+    }
+
+    pub fn chunk_production_key(&self) -> ChunkProductionKey {
+        match self {
+            Self::V1(accesses) => accesses.inner.next_chunk.clone().into(),
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, BorshSerialize, BorshDeserialize, ProtocolSchema)]
+pub struct ChunkContractAccessesV1 {
+    inner: ChunkContractAccessesInner,
+    /// Signature of the inner, signed by the chunk producer of the next chunk.
+    signature: Signature,
+}
+
+impl ChunkContractAccessesV1 {
+    fn new(
+        next_chunk: ChunkProductionKey,
+        contracts: Vec<CodeHash>,
+        signer: &ValidatorSigner,
+    ) -> Self {
+        let inner = ChunkContractAccessesInner::new(next_chunk, contracts);
+        let signature = signer.sign_chunk_contract_accesses(&inner);
+        Self { inner, signature }
+    }
+}
+
+/// Identifies a chunk by the epoch, block, and shard in which it was produced.
+#[derive(Debug, Clone, PartialEq, Eq, BorshSerialize, BorshDeserialize, ProtocolSchema)]
+struct ChunkMetadata {
```

Comment:
why do we need a separate struct? `ChunkProductionKey` was introduced exactly for that

Reply:
Discussed offline, and I will make `ChunkProductionKey` borsh-serializable and protocol schema and use it.

Reply:
done